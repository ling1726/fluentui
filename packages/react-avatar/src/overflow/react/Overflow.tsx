import * as React from 'react';
import { useMergedRefs } from '@fluentui/react-utilities';
import { OverflowGroupState, OnUpdateOverflow, OverflowEventPayload } from '../native/overflowManager';
import { OverflowContext } from './overflowContext';
import { useOverflowContainer, UseOverflowContainerOptions } from './useOverflowContainer';

export type OverflowChildProps = React.HTMLAttributes<HTMLElement> & {
  ref?: React.Ref<never>;
};

export type OverflowProps = UseOverflowContainerOptions & {
  children: React.ReactElement;
};

export const Overflow = React.forwardRef((props: OverflowProps, ref) => {
  const [hasOverflow, setHasOverflow] = React.useState(false);
  const [itemVisibility, setItemVisibility] = React.useState<Record<string, boolean>>({});
  const [groupVisibility, setGroupVisibility] = React.useState<Record<string, OverflowGroupState>>({});

  const updateItemVisibility: OnUpdateOverflow = (data: OverflowEventPayload) => {
    setHasOverflow(() => data.invisibleItems.length > 0);
    setItemVisibility(() => {
      const newState: Record<string, boolean> = {};
      data.visibleItems.forEach(x => (newState[x.id] = true));
      data.invisibleItems.forEach(x => (newState[x.id] = false));
      return newState;
    });
    setGroupVisibility(groupVisibility);
  };

  const { containerRef, registerItem, updateOverflow } = useOverflowContainer(updateItemVisibility, {
    ...props,
  });

  const mergedRef = useMergedRefs(containerRef, ref);

  const children = React.cloneElement(React.Children.only(props.children), { ref: mergedRef });

  return (
    <OverflowContext.Provider
      value={{
        itemVisibility,
        groupVisibility,
        hasOverflow,
        registerItem,
        updateOverflow,
      }}
    >
      {children}
    </OverflowContext.Provider>
  );
});
