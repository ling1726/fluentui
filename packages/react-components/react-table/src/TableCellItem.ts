export * from './components/TableCellItem/index';
