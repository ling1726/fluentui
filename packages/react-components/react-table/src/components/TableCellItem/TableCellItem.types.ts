import type { ComponentProps, ComponentState, Slot } from '@fluentui/react-utilities';

export type TableCellItemSlots = {
  root: Slot<'div'>;

  media?: Slot<'span'>;
};

/**
 * TableCellItem Props
 */
export type TableCellItemProps = ComponentProps<TableCellItemSlots> & {};

/**
 * State used in rendering TableCellItem
 */
export type TableCellItemState = ComponentState<TableCellItemSlots>;
