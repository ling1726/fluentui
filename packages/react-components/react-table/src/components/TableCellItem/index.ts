export * from './TableCellItem';
export * from './TableCellItem.types';
export * from './renderTableCellItem';
export * from './useTableCellItem';
export * from './useTableCellItemStyles';
