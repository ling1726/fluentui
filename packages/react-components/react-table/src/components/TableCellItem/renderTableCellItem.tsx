import * as React from 'react';
import { getSlots } from '@fluentui/react-utilities';
import type { TableCellItemState, TableCellItemSlots } from './TableCellItem.types';

/**
 * Render the final JSX of TableCellItem
 */
export const renderTableCellItem_unstable = (state: TableCellItemState) => {
  const { slots, slotProps } = getSlots<TableCellItemSlots>(state);

  return (
    <slots.root {...slotProps.root}>
      {slots.media && <slots.media {...slotProps.media} />} {slotProps.root.children}
    </slots.root>
  );
};
