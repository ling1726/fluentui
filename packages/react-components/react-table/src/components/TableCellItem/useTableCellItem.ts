import * as React from 'react';
import { getNativeElementProps, resolveShorthand } from '@fluentui/react-utilities';
import type { TableCellItemProps, TableCellItemState } from './TableCellItem.types';

/**
 * Create the state required to render TableCellItem.
 *
 * The returned state can be modified with hooks such as useTableCellItemStyles_unstable,
 * before being passed to renderTableCellItem_unstable.
 *
 * @param props - props from this instance of TableCellItem
 * @param ref - reference to root HTMLElement of TableCellItem
 */
export const useTableCellItem_unstable = (
  props: TableCellItemProps,
  ref: React.Ref<HTMLElement>,
): TableCellItemState => {
  return {
    components: {
      root: 'div',
      media: 'span',
    },
    root: getNativeElementProps('div', {
      ref,
      ...props,
    }),
    media: resolveShorthand(props.media),
  };
};
