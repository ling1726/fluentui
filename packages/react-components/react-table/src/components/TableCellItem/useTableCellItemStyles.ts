import { makeStyles, mergeClasses, shorthands } from '@griffel/react';
import { tokens } from '@fluentui/react-theme';
import type { TableCellItemSlots, TableCellItemState } from './TableCellItem.types';
import type { SlotClassNames } from '@fluentui/react-utilities';

export const tableCellItemClassNames: SlotClassNames<TableCellItemSlots> = {
  root: 'fui-TableCellItem',
  media: 'fui-TableCellItem__media',
};

/**
 * Styles for the root slot
 */
const useStyles = makeStyles({
  root: {
    display: 'flex',
    alignItems: 'center',
    ...shorthands.gap(tokens.spacingHorizontalS),
    ...shorthands.flex(1, 1, '0px'),
  },

  media: {
    display: 'flex',
    alignItems: 'center',
  },

  // TODO add additional classes for different states and/or slots
});

/**
 * Apply styling to the TableCellItem slots based on the state
 */
export const useTableCellItemStyles_unstable = (state: TableCellItemState): TableCellItemState => {
  const styles = useStyles();
  state.root.className = mergeClasses(tableCellItemClassNames.root, styles.root, state.root.className);
  if (state.media) {
    state.media.className = mergeClasses(tableCellItemClassNames.media, styles.media, state.media.className);
  }

  return state;
};
