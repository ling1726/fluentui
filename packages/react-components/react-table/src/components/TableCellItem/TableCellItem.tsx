import * as React from 'react';
import { useTableCellItem_unstable } from './useTableCellItem';
import { renderTableCellItem_unstable } from './renderTableCellItem';
import { useTableCellItemStyles_unstable } from './useTableCellItemStyles';
import type { TableCellItemProps } from './TableCellItem.types';
import type { ForwardRefComponent } from '@fluentui/react-utilities';

/**
 * TableCellItem component - TODO: add more docs
 */
export const TableCellItem: ForwardRefComponent<TableCellItemProps> = React.forwardRef((props, ref) => {
  const state = useTableCellItem_unstable(props, ref);

  useTableCellItemStyles_unstable(state);
  return renderTableCellItem_unstable(state);
});

TableCellItem.displayName = 'TableCellItem';
