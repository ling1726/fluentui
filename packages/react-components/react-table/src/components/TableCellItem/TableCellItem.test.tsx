import * as React from 'react';
import { render } from '@testing-library/react';
import { TableCellItem } from './TableCellItem';
import { isConformant } from '../../common/isConformant';

describe('TableCellItem', () => {
  isConformant({
    Component: TableCellItem,
    displayName: 'TableCellItem',
  });

  // TODO add more tests here, and create visual regression tests in /apps/vr-tests

  it('renders a default state', () => {
    const result = render(<TableCellItem>Default TableCellItem</TableCellItem>);
    expect(result.container).toMatchSnapshot();
  });
});
