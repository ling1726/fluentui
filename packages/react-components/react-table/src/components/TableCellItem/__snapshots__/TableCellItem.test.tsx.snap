// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableCellItem renders a default state 1`] = `
<div>
  <div
    class="fui-TableCellItem"
  >
     
    Default TableCellItem
  </div>
</div>
`;
